core/src/main/scala/kafka/zk/KafkaZkClient.scala

```
@@ -183,7 +190,15 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
         case Code.OK =>
           val updatedLeaderAndIsr = leaderAndIsrs(partition).withZkVersion(setDataResponse.stat.getVersion)
           successfulUpdates.put(partition, updatedLeaderAndIsr)
-        case Code.BADVERSION => updatesToRetry += partition
+        case Code.BADVERSION =>
+          val checkSucceed = setDataResponse.checkResult match {
+            case Some(checkResult) => checkResult.resultCode == Code.OK
+            case None => false
+          }
+          if (checkSucceed)
+            updatesToRetry += partition
+          else
+            failed.put(partition, setDataResponse.resultException.get)
```

In general we want to the exception returned to the caller to uniquely identify the problem. But here we can output the BadVersionException if either the controller epoch zkversion is bad or the data znode zkversion is bad. It maybe confusing.

It is probably simpler to create a new exception, e.g. `ControllerEpochZkVersionMismatchException` and do the following before checking `setDataResponse.resultCode`:

```
if (setDataResponse.zkVersionCheckResultCode != Code.OK)
  failed.put(partition, new ControllerEpochZkVersionMismatchException())
```

---

Thanks for the suggestion. It is simpler and cleaner this way.

Btw, I think we can just use `ControllerMoveException()` in this case. 

Also, related to your comment on my last commit, instead of putting the exception in some data structures, can we just simply throw the exception when the check fails? In this case, we can skip processing unnecessary controller event until we hit `ControllerChange` event. To optimize it further, we can also catch 'ContollerMoveException' explicitly in the `ControllerEventThread` and let the controller resigns immediately.